webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,249 @@
+  ## Creates an encoded Binding Request
+  ##
+  result = StunMessage(msgType: StunBindingRequest)
+  self.rng[].generate(result.transactionId)
+
+  let username = self.usernameProvider()
+  if username != "":
+    result.attributes.add(UsernameAttribute.encode(username))
+
+  if self.iceControlling:
+    result.attributes.add(IceControlling.encode(self.iceTiebreaker))
+  else:
+    result.attributes.add(IceControlled.encode(self.iceTiebreaker))
+  result.attributes.add(Priority.encode(self.calculatePriority()))
+
+proc checkForError*(self: StunConn, msg: StunMessage): Option[StunMessage] =
+  # Check for error from a BindingRequest message.
+  # Returns an option with some BindingErrorResponse if there is an error.
+  # Returns none otherwise.
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-10.1.2
+  var res = StunMessage(msgType: StunBindingErrorResponse,
+                        transactionId: msg.transactionId)
+  if msg.getAttribute(MessageIntegrity).isNone() or
+     msg.getAttribute(UsernameAttribute).isNone():
+    res.attributes.add(ErrorCode.encode(ECBadRequest))
+    return some(res)
+
+  let usernameAttr = msg.getAttribute(UsernameAttribute).get()
+  if not self.usernameChecker(usernameAttr.username):
+    res.attributes.add(ErrorCode.encode(ECUnauthorized))
+    return some(res)
+
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-15.9
+  var unknownAttr: seq[uint16]
+  for attr in msg.attributes:
+    let typ = attr.attributeType
+    if typ.isRequired() and typ notin StunAttributeEnum:
+      unknownAttr.add(typ)
+  if unknownAttr.len() > 0:
+    res.attributes.add(ErrorCode.encode(ECUnknownAttribute))
+    res.attributes.add(UnknownAttributes.encode(unknownAttr))
+    return some(res)
+
+  return none(StunMessage)
+
+proc isFingerprintValid*(msg: StunMessage): bool =
+  # Returns true if Fingerprint is missing or if it's valid.
+  # Returns false otherwise.
+  let fingerprint = msg.getAttribute(Fingerprint)
+  if fingerprint.isNone():
+    return true
+  if msg.attributes[^1].attributeType != AttrFingerprint:
+    # Fingerprint should always be the last attribute.
+    return false
+  let
+    copyWithoutFingerprint = StunMessage(
+      msgType: msg.msgType,
+      transactionId: msg.transactionId,
+      attributes: msg.attributes[0 ..< ^1]
+    )
+    encodedCopy = copyWithoutFingerprint.encode(@[])
+  return fingerprint == StunMessage.decode(encodedCopy).getAttribute(Fingerprint)
+
+# - Stun Messages Handler -
+
+proc stunMessageHandler(self: StunConn) {.async: (raises: [CancelledError]).} =
+  # Read indefinitely Stun messages from stunMsgs queue.
+  # Sends a BindingResponse or BindingResponseError after receiving a BindingRequest.
+  while true:
+    let message = await self.stunMsgs.popFirst()
+    try:
+      let decoded = StunMessage.decode(await self.stunMsgs.popFirst())
+      if not decoded.isFingerprintValid():
+        # Fingerprint is invalid, the StunMessage received might be a false positive.
```

Comment:
What does "might be a false positive" mean? Not a Stun message? Why is it moved to the `dataRecv` queue?

Reply:
Basically, there's a first check with the raw data (without decoding) where we check different things (the size of the message, the presence of the magic stun cookie etc...). When the incoming message is sorted, we can decode it. If something is wrong with the Fingerprint, it can means that it is, in fact, not a Stun Message but a message for another protocol. And the RFC specifies this by saying :
```
   The FINGERPRINT attribute can aid in distinguishing STUN packets from
   packets of other protocols.  See [Section 7](https://datatracker.ietf.org/doc/html/rfc8489#section-7).```

Reply:
Can you briefly describe it in the comment and add the link? Why is it moved to the dataRecv queue?